lib/Dialect/TritonGPU/Transforms/Pipeliner/MatmulLoopPipeline.cpp

```diff
@@ -811,46 +801,57 @@ minWaitNumberForExtract(ttg::ExtractSliceOp extractOp) {
       int min2 = minOverHistories(prevVal, yieldOp, thisHistorySum);
       return std::min(std::min(min1, min2), minCommitNumber);
     }
-    // Failed to track, return 1 conservatively.
-    return 1;
+    // Failed to track, return 0 conservatively.
+    return 0;
   };
 
-  int minCommits = minOverHistories(extractOp.getOperand(0), extractOp, 0);
-  if (minCommits == 0)
-    llvm::report_fatal_error("No commits between insert and extract!");
-  return minCommits - 1;
+  if (waitOp->getNumOperands() != 1)
+    return 0;
+  int minCommits = minOverHistories(waitOp->getOperand(0), waitOp, 0);
+  return minCommits;
 }
 
-/// Insert wait ops after the extract_slice ops.
-void mlir::triton::insertWaits(ModuleOp module) {
-  module.walk([&](ttg::ExtractSliceOp firstExtractOp) {
-    if (!firstExtractOp->hasAttr(kNeedWaitAttrName))
-      return;
-
-    Operation *extractOp = firstExtractOp;
-    ttg::ExtractSliceOp lastExtractOp = firstExtractOp;
-
-    // If there is no meaningful work between the extracts, don't insert
-    // multiple waits. Insert just one wait per group of extracts.
-    std::optional<int> minWaitNumber = std::nullopt;
-    while (extractOp) {
-      lastExtractOp = cast<ttg::ExtractSliceOp>(extractOp);
-      std::optional<int> currMin = minWaitNumberForExtract(lastExtractOp);
-      if (currMin.has_value())
-        minWaitNumber =
-            std::min(minWaitNumber.value_or(INT_MAX), currMin.value());
-
-      extractOp->removeAttr(kNeedWaitAttrName);
-      extractOp = dyn_cast<ttg::ExtractSliceOp>(extractOp->getNextNode());
+// Look for consecutive wait ops and combine them into a single wait op.
+static void
+combineRedundantWaitOps(llvm::SmallSetVector<ttg::AsyncWaitOp, 8> &waitOps) {
+  llvm::SmallSetVector<ttg::AsyncWaitOp, 8> toDelete;
+  for (auto waitOp : waitOps) {
+    if (toDelete.count(waitOp))
+      continue;
+    SmallVector<ttg::AsyncWaitOp> waitGroup = {waitOp};
+    SmallVector<Value> depTokens;
+    unsigned minWaitNumber = waitOp.getNum();
+    Operation *next = waitOp->getNextNode();
+    while (next && isa<ttg::SubviewOp, ttg::AsyncWaitOp>(next)) {
```

Comment:
Why `SubviewOp` is special?

Reply:
it's kind of ad hoc and could technically be any pure op but didn't want to go too generic. The current pipeliner schedule generates subview followed by wait so I handle this special case, if we start doing more complex stuff we might want a smarter logic.

Reply:
> The current pipeliner schedule generates subview followed by wait so I handle this special case

I see, I thought many types of ops can be skipped between two async wait ops.

Reply:
correct